### ISSUE TITLE
Boxplots - bugs and suggestions

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/Instat#1940]
This has come from our work with the Lesotho team.  First this message is on boxplots, which has thrown up some possibly general issues on the graphics at the same time:

First is that we hit 3 problems with the boxplots.

They are when you use multiple variables. We were constructing boxplots for both maxtemp and mintemp together. We wanted the x to be the factor called years.
1. This isn't currently possible. It makes the x-variable into the variables, and the years is the second factor. It would be good to be able to switch them round. Perhaps a checkbox is the simplest way?
2. When we added facets as months it crashes.
3. So I used the stacked data instead, where mintemp and maxtemp are in the same column. Now I first filtered out the rain variable. Then I did the boxplot. It ignored my filter. (Boxplots doesn't always ignore filters, but I think there could be a problem when a complete level is ignored.)  But see point 3 below.

Now some further points:
1. We managed to make the width of the boxplots proportional to the number of observations - which is very useful.  But we needed to go to **Boxplot Options**, then change the tab, and then click on the **varwidth** option.  Then it opens a pull-down box with just True/False which isn't needed.  When we check then it is true! 
2.  But this brings a more general point.  When we are doing boxplots there will be some options that we would often like.  Could they go on the main dialogue - as well?  This is one of them.  A second could be the coord-flip for horizontal boxplots.
3.  On a positive note I have been trying the boxplots with more filtering options and they work well.  I am a bit confused on what has been happening in point 3 above.
4.  I would very much like to be able to locate points (possibly just outliers) or interactively on the boxplots.  Interactive is a big step.  But I wonder about adding information from another column to the outliers, or even labelling with the exact value.  In stack overflow here is an answer, but it seems a bit complicated: http://stackoverflow.com/questions/33524669/labeling-outliers-of-boxplots-in-r 
   Maybe http://rrubyperlundich.blogspot.co.uk/2015/07/1-label-outlier-in-ggplot2-boxplot.html 
   is better?

Roger

**Comment by maxwellfundi** [comment]
@rdstern  great, we will look at this . Could you please send us the data you used. It might be quite helpful to test what you were doing

**Comment by rdstern** [comment]
The data we used are being sent separately.  github doesn't support rds files, and I hit a bug trying to export to Excel.

While on this topic I should mention how well boxplots are working.  They have really helped the analyses from Lesotho.  Of course I have some more suggestions!

We are using year as a factor.  This is the x-axis and that will be a common aspect.  The levels are 1971, 1972, ... 2015 and of course they overlap.  It would be better if they could be displayed vertically, rather than horizontally.  I assume this is all part of the themes that are still to come.  But it would be great if common features (like this?) were on the main dialogue.  That would also give more of a point to having the "Specific" menu.

A second possible feature for the main boxplot dialogue is that it would often be good to add the jittered points to the graph, i.e. another layer.  Again I suggest this would be common enough that this might be a checkbox on the main boxplot dialogue.  

And a third option might be to allow violin plots as well, or instead of boxplots.  See http://www.sthda.com/english/wiki/ggplot2-violin-plot-quick-start-guide-r-software-and-data-visualization 

Roger

**Comment by FrancoisJRenaud** [comment]
Hi, a few comments and questions before implementing/exploring some changes. 

About the 3 problems : 

1. Do we agree on the check box for flipping the mapping of the "variables" factor and additional factor from x - fill resp. to fill - x. How should I name that chk box ? 
Alternatively, I could add a button "flip factors" that automatically exchanges the content of the two different receivers ? I think that's feasible as well.
2. will have a look at this.
3. I don't have a clear picture of what happens with the interaction between LayerOptions and BoxPlot in multiple receiver mode. This is the error that I get when I went to LayerOptions and come back. 
![boxplotbug](https://cloud.githubusercontent.com/assets/22272110/19922261/e5c3fd50-a0f2-11e6-807b-3162905f9c79.png)
here we can see a successfully run line before going to layer options and then the bugging line. The problem seems to come from the the method measure.vars(...) that is not called anymore when coming back from the options. 
![boxplotbug2](https://cloud.githubusercontent.com/assets/22272110/19923323/d3c1dd98-a0f7-11e6-8438-2de088c18f6f.png)

I am mentioning this here because we can see the Use_current_filter=False appearing everywhere. Is there a reason for it to be there ? If I m not mistaken all these specific graph dialogs run a line like : 
survey_temp <- InstatDataObject$get_data_frame(use_current_filter=FALSE, data_name="survey")
before running ggplot.
 I guess that's also why the filter is ignored. Would like to discuss this with @dannyparsons maybe. 

About the other remarks: 
1. I could try to not have that True or False combobox, but I don't think there is an easy clean way to do this. (discussed with @maxwellfundi, don't know if you have a suggestion @dannyparsons) Maybe could add this "just tick" type of parameters in the list of parameter types we have, but I have the feeling that it is a lot of efforts for something that's not really crucial. What do you think @rdstern ?
2. So I will add the varwidth parameter on the dlgBoxPlot. There it can just be a tick of course. 
I think the flip coord has already been implemented ? the chk box Horizontal ? 
3. @rdstern  When you mention filtering options, you are just working with a data frame on which a filter has been applied right ? My plots always seem to ignore filters but not quite sure. In which particular cases did your filters apply ? 
4. I will leave this one for later.

**Comment by dannyparsons** [comment]
I think changing the mapping of `variable` will take some more thought. How will this work on other dialogs? The factors on the main dialog for boxplot corresponds to x and fill - but there's also things like colour. So maybe `variable` should also be able to map to colour and other aes as well?

When you get errors like that - the error message is not usually sensible for graphs. Take the R code and run it in RStudio to get much better error messages. Also, the commands are always displayed in the output window, regardless of whether they give an error or not, so that isn't any indication of working/not working.

use_current_filter = FALSE shouldn't be in there. I'm not sure why it is but I've fixed the code for it, so when I merge that it should work again and use the filter for graphs.

We can discuss about the TRUE/FALSE vs checkbox here as well, it's the way we implemented that control to make it general for enough for other types of parameter. I think if there is a checkbox for it as well on the main dialog, that will be good and avoid confusion for the typical user.

Remember when adding options on the main dialogs, that you also test that they link properly with the subdialog. So if I check varwidth on the main dialog, this should be reflected in the parameters list on the layer subdialog.
